File: ParralaxController/src/tankStatus.c
#include <TankStatus/TankStatus.h>
#include <string.h> // Required for memcpy

/**
 * Serializes the TankStatus struct into a byte buffer for USART streaming.
 */
void constructTankStatus(struct TankStatus *self) {
  self->driveLeft = 0;
  self->driveRight = 0;
  self->eulerX = 0;
  self->eulerY = 0;
  self->eulerZ = 90;
  self->changeFlag = 0;
  // self->isLocked = 0;
}

void makeByteTankStatus(unsigned char *buffer, int byteLength,
                        struct TankStatus *ts) {
  // Safety check: ensure buffer is large enough
  if (byteLength < TANKSTATUS_PACKET_LENGTH || buffer == NULL || ts == NULL) {
    return;
  }

  int offset = 0;

  memcpy(buffer + offset, (const void *)&ts->driveLeft, sizeof(ts->driveLeft));
  offset += sizeof(ts->driveLeft);

  memcpy(buffer + offset, (const void *)&ts->driveRight,
         sizeof(ts->driveRight));
  offset += sizeof(ts->driveRight);

  // Padding REMOVED. Packing shoulder-to-shoulder!

  memcpy(buffer + offset, (const void *)&ts->eulerX, sizeof(ts->eulerX));
  offset += sizeof(ts->eulerX);

  memcpy(buffer + offset, (const void *)&ts->eulerY, sizeof(ts->eulerY));
  offset += sizeof(ts->eulerY);

  memcpy(buffer + offset, (const void *)&ts->eulerZ, sizeof(ts->eulerZ));
}

void readByteTankStatus(unsigned char *buffer, int byteLength,
                        struct TankStatus *ts) {
  // Safety check
  if (byteLength < TANKSTATUS_PACKET_LENGTH || buffer == NULL || ts == NULL) {
    return;
  }

  int offset = 0;

  memcpy((void *)&ts->driveLeft, buffer + offset, sizeof(ts->driveLeft));
  offset += sizeof(ts->driveLeft);

  memcpy((void *)&ts->driveRight, buffer + offset, sizeof(ts->driveRight));
  offset += sizeof(ts->driveRight);

  // Padding REMOVED.

  memcpy((void *)&ts->eulerX, buffer + offset, sizeof(ts->eulerX));
  offset += sizeof(ts->eulerX);

  memcpy((void *)&ts->eulerY, buffer + offset, sizeof(ts->eulerY));
  offset += sizeof(ts->eulerY);

  memcpy((void *)&ts->eulerZ, buffer + offset, sizeof(ts->eulerZ));
}
/*
float getEulerfloat(short data) {
  // data should be a byte FF where first is the char, second half is the dec
  // float dec = data & 0xFF >> 8; // decimal portion
  float dec = (data & 0xFF) / 256.0f;
  float dataOut = data >> 8; // integer portion
  return dataOut + dec;
}

int makeEulerInt(float data);
*/

File: TankBotWebUI/src/tankstatus.cs
using System;
using System.Runtime.InteropServices;

// Matches the 'struct TankStatus' from C
[StructLayout(LayoutKind.Sequential)]
public struct TankStatus
{
    public byte driveLeft;
    public byte driveRight;
    public short eulerX;
    public short eulerY;
    public short eulerZ;
    public byte changeFlag;
}

public static class TankStatusNative
{
    public const int PacketLength = 16;

    public static void constructTankStatus(ref TankStatus self)
    {
        self.driveLeft = 0;
        self.driveRight = 0;
        self.eulerX = 0;
        self.eulerY = 0;
        self.eulerZ = 90;
        self.changeFlag = 0;
    }

    public static void makeByteTankStatus(byte[] buffer, int byteLength, ref TankStatus ts)
    {
        if (byteLength < PacketLength || buffer == null || buffer.Length < PacketLength) return;

        buffer[0] = ts.driveLeft;
        buffer[1] = ts.driveRight;
        
        // Offset 2 and 3 are padding, skip them
        
        byte[] ex = BitConverter.GetBytes(ts.eulerX);
        buffer[4] = ex[0];
        buffer[5] = ex[1];

        byte[] ey = BitConverter.GetBytes(ts.eulerY);
        buffer[6] = ey[0];
        buffer[7] = ey[1];

        byte[] ez = BitConverter.GetBytes(ts.eulerZ);
        buffer[8] = ez[0];
        buffer[9] = ez[1];
    }

    public static void readByteTankStatus(byte[] buffer, int byteLength, ref TankStatus ts)
    {
        if (byteLength < PacketLength || buffer == null || buffer.Length < PacketLength) return;

        ts.driveLeft = buffer[0];
        ts.driveRight = buffer[1];
        
        // Offset 2 and 3 are padding, skip them
        
        ts.eulerX = BitConverter.ToInt16(buffer, 4);
        ts.eulerY = BitConverter.ToInt16(buffer, 6);
        ts.eulerZ = BitConverter.ToInt16(buffer, 8);
    }

    public static int get_tankstatus_packet_length()
    {
        return PacketLength;
    }

    public static float getDecodedShortToFloat(short data)
    {
        return (float)data / 256.0f;
    }

    public static short getEncodedFloatToShort(float reading)
    {
        return (short)(reading * 256.0f);
    }
}
